File: src/mocks/handlers/alertas.ts
import { http, HttpResponse } from 'msw';
import { mockAlertas } from '../data/alertas';
import { getUserFromRequest } from '../utils';

export const alertasHandlers = [
  http.get('*/api/alertas', ({ request }) => {
    const url = new URL(request.url);
    const severidad = url.searchParams.get('severidad');
    const tipo = url.searchParams.get('tipo');
    const silenciada = url.searchParams.get('silenciada');
    const resuelta = url.searchParams.get('resuelta');
    const tiendaId = url.searchParams.get('tienda_id');
    const search = url.searchParams.get('search')?.toLowerCase();

    let filtered = [...mockAlertas];

    const user = getUserFromRequest(request);
    if (user?.rol === 'OPERATOR' && user.tiendas_asignadas) {
      filtered = filtered.filter((a) => user.tiendas_asignadas!.includes(a.tienda_id));
    }

    if (search) {
      filtered = filtered.filter(
        (a) =>
          a.mensaje.toLowerCase().includes(search) ||
          (a.tienda_nombre && a.tienda_nombre.toLowerCase().includes(search)) ||
          (a.tramite_nombre && a.tramite_nombre.toLowerCase().includes(search)),
      );
    }

    if (severidad) filtered = filtered.filter((a) => a.severidad === severidad);
    if (tipo) filtered = filtered.filter((a) => a.tipo === tipo);
    if (silenciada !== null && silenciada !== undefined) {
      filtered = filtered.filter((a) => a.silenciada === (silenciada === 'true'));
    }
    if (resuelta !== null && resuelta !== undefined) {
      filtered = filtered.filter((a) => !!a.resuelta === (resuelta === 'true'));
    }
    if (tiendaId) filtered = filtered.filter((a) => a.tienda_id === tiendaId);

    return HttpResponse.json(filtered);
  }),

  http.post('*/api/alertas/:id/silenciar', async ({ params, request }) => {
    const body = (await request.json()) as { duracion_dias: number; nota?: string };
    const index = mockAlertas.findIndex((a) => a.id === params.id);
    if (index === -1) {
      return HttpResponse.json({ detail: 'Alerta no encontrada' }, { status: 404 });
    }

    mockAlertas[index] = {
      ...mockAlertas[index],
      silenciada: true,
      silenciada_hasta: new Date(Date.now() + body.duracion_dias * 86400000).toISOString(),
      silenciada_por: 'usr-001',
    };

    return HttpResponse.json(mockAlertas[index]);
  }),

  http.post('*/api/alertas/:id/resolver', async ({ params }) => {
    const index = mockAlertas.findIndex((a) => a.id === params.id);
    if (index === -1) return HttpResponse.json({ detail: 'Alerta no encontrada' }, { status: 404 });

    mockAlertas[index] = {
      ...mockAlertas[index],
      silenciada: false,
      resuelta: true,
      fecha_resolucion: new Date().toISOString(),
      resuelta_por: 'usr-001',
    };
    return HttpResponse.json(mockAlertas[index]);
  }),

  http.post('*/api/alertas/:id/reactivar', async ({ params }) => {
    const index = mockAlertas.findIndex((a) => a.id === params.id);
    if (index === -1) return HttpResponse.json({ detail: 'Alerta no encontrada' }, { status: 404 });

    mockAlertas[index] = {
      ...mockAlertas[index],
      silenciada: false,
      silenciada_hasta: undefined,
      silenciada_por: undefined,
    };
    return HttpResponse.json(mockAlertas[index]);
  }),

  http.post('*/api/alertas/:id/notificar/:canal', async ({ params }) => {
    const index = mockAlertas.findIndex((a) => a.id === params.id);
    if (index === -1) return HttpResponse.json({ detail: 'Alerta no encontrada' }, { status: 404 });

    await new Promise((r) => setTimeout(r, 800));

    const canal = params.canal as 'email' | 'whatsapp';
    mockAlertas[index] = {
      ...mockAlertas[index],
      notificaciones_enviadas: {
        ...(mockAlertas[index].notificaciones_enviadas || { email: false, whatsapp: false }),
        [canal]: true,
      },
    };
    return HttpResponse.json(mockAlertas[index]);
  }),

  // Count of active critical alerts (for sidebar badge)
  http.get('*/api/alertas/count', ({ request }) => {
    let alertas = [...mockAlertas];
    const user = getUserFromRequest(request);

    if (user?.rol === 'OPERATOR' && user.tiendas_asignadas) {
      alertas = alertas.filter((a) => user.tiendas_asignadas!.includes(a.tienda_id));
    }

    const criticalCount = alertas.filter(
      (a) => !a.silenciada && !a.resuelta && a.severidad === 'critical',
    ).length;
    return HttpResponse.json({ count: criticalCount });
  }),
];

File: backend/app/api/alertas.py
from fastapi import APIRouter

from app.api.deps import CurrentUser, DbSession
from app.schemas.alerta import AlertaCountOut, AlertaOut, AlertaSilenciarRequest
from app.services import alerta_service

router = APIRouter()


def _serialize_alerta(a) -> dict:
    return {
        "id": a.id,
        "tipo": a.tipo,
        "severidad": a.severidad,
        "tienda_id": a.tienda_id,
        "tienda_nombre": getattr(a.tienda, "nombre", None)
        if hasattr(a, "tienda")
        else None,
        "tramite_id": a.tramite_id,
        "tramite_nombre": getattr(a.tramite, "nombre", None)
        if hasattr(a, "tramite")
        else None,
        "documento_id": a.documento_id,
        "mensaje": a.mensaje,
        "fecha_generacion": a.fecha_generacion.isoformat(),
        "silenciada": a.silenciada,
        "silenciada_hasta": a.silenciada_hasta.isoformat()
        if a.silenciada_hasta
        else None,
        "silenciada_por": a.silenciada_por,
        "resuelta": a.resuelta,
        "fecha_resolucion": a.fecha_resolucion.isoformat()
        if a.fecha_resolucion
        else None,
        "resuelta_por": a.resuelta_por,
        "notificaciones_enviadas": a.notificaciones_enviadas,
    }


@router.get("", response_model=list[AlertaOut])
async def list_alertas(
    db: DbSession,
    current_user: CurrentUser,
    severidad: str | None = None,
    tipo: str | None = None,
    silenciada: bool | None = None,
    resuelta: bool | None = None,
    tienda_id: str | None = None,
    search: str | None = None,
):
    items = await alerta_service.get_many(
        db,
        severidad=severidad,
        tipo=tipo,
        silenciada=silenciada,
        resuelta=resuelta,
        tienda_id=tienda_id,
        search=search,
        current_user=current_user,
    )
    return [_serialize_alerta(a) for a in items]


@router.get("/count/critical", response_model=AlertaCountOut)
async def count_critical(db: DbSession, current_user: CurrentUser):
    count = await alerta_service.count_active_critical(db, current_user=current_user)
    return {"count": count}


@router.post("/{id}/silenciar", response_model=AlertaOut)
async def silenciar(
    db: DbSession, id: str, data: AlertaSilenciarRequest, current_user: CurrentUser
):
    a = await alerta_service.silenciar(
        db, id, duracion_dias=data.duracion_dias, nota=data.nota, actor=current_user
    )
    return _serialize_alerta(a)


@router.post("/{id}/resolver", response_model=AlertaOut)
async def resolver(db: DbSession, id: str, current_user: CurrentUser):
    a = await alerta_service.resolver(db, id, actor=current_user)
    return _serialize_alerta(a)


@router.post("/{id}/reactivar", response_model=AlertaOut)
async def reactivar(db: DbSession, id: str, current_user: CurrentUser):
    a = await alerta_service.reactivar(db, id, actor=current_user)
    return _serialize_alerta(a)


@router.post("/{id}/notificar/{canal}", response_model=AlertaOut)
async def notificar(db: DbSession, id: str, canal: str, current_user: CurrentUser):
    a = await alerta_service.notificar(db, id, canal=canal, actor=current_user)
    return _serialize_alerta(a)
